### PR TITLE
Change unbond duration from eras to time left

### PR DESCRIPTION
**Pull request by TingALin**
This PR links to issue #520 

- Create new hook, useErasToTimeleft, where you can pass a number of eras into the hook and get the corresponding duration in seconds back.
- library/Hooks/useTimeLeft.tsx: Evolve timeleftAsString(): string to timeLeftAsString(to?: Date):string. Handle the current era (toRef.current) if no to is provided as an argument.
- Have an option to either do “full” format (hours, minutes) or not (hrs, mins). Full can be a Boolean argument that defaults to false
- Get rid of “0 hours” /“0 minutes”if hrs/mins is zero

### CODE REVIEW

File: src/modals/UnlockChunks/Overview.tsx [hunk 1]
```diff
@@ -25,6 +28,14 @@ export const Overview = forwardRef(
     const { isFastUnstaking } = useUnstaking();
     const { t } = useTranslation('modals');
 
+    const { timeleft: eraTimeLeft } = useEraTimeLeft();
+    const { fromNow, timeleftAsString } = useTimeLeft({
+      refreshInterval: 60,
+      refreshCallback: () => eraTimeLeft,
+    });
+    const { durationSeconds } = useErasToTimeleft(bondDuration);
```
Comment (rossbulat) [inline src/modals/UnlockChunks/Overview.tsx:36]: When I tested the modal the text flickers between `12 hours` and `11 hours 59 mins` every so often. 

This value does not need to be a live, it is always 12 hours that needs to be statically converted from eras to time.

File: src/library/Hooks/useTimeLeft/index.tsx [hunk 2]
```diff
@@ -184,21 +184,31 @@ export const useTimeLeft = ({
   }, []);
 
   // format the duration as a string.
-  const timeleftAsString = (() => {
-    const { days, hours, minutes, seconds } = getDuration(toRef.current);
+  const timeleftAsString = (toDate?: Date, fullFormat?: boolean) => {
```
Comment (rossbulat) [inline src/library/Hooks/useTimeLeft/index.tsx:187]: change `fullFormat` to just `full`.

File: src/library/Hooks/useErasToTimeleft/index.tsx [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+  // store the duration of an era in block numbers.
+  const eraDurationBlocks = epochDuration * sessionsPerEra;
+
+  // estimate the duration of the era in seconds
+  const eraDuration = eraDurationBlocks * expectedBlockTime * 0.001;
+
+  const durationSeconds = eras * eraDuration;
+
+  return { durationSeconds };
```
Comment (rossbulat) [inline src/library/Hooks/useErasToTimeleft/index.tsx:18]: Do not make this hook calculate the timeleft immediately after initialisation. 

Instead, let it provide a function that can be imported to a component and called when needed.

```suggestion

const erasToTimeLeft = (eras: number) => {
  const { consts } = useApi();
  const { epochDuration, expectedBlockTime, sessionsPerEra } = consts;
  // store the duration of an era in block numbers.
  const eraDurationBlocks = epochDuration * sessionsPerEra;
  // estimate the duration of the era in seconds
  const eraDuration = eraDurationBlocks * expectedBlockTime * 0.001;
  const durationSeconds = eras * eraDuration;
  return durationSeconds;
  }
  
  return {
     erasToTimeLeft
  };
```

File: src/modals/UnlockChunks/Overview.tsx [hunk 4]
```diff
@@ -25,6 +28,14 @@ export const Overview = forwardRef(
     const { isFastUnstaking } = useUnstaking();
     const { t } = useTranslation('modals');
 
+    const { timeleft: eraTimeLeft } = useEraTimeLeft();
```
Comment (rossbulat) [inline src/modals/UnlockChunks/Overview.tsx:31]: This is not needed. We are not working with a live era count down.

File: src/modals/UnlockChunks/Overview.tsx [hunk 5]
```diff
@@ -25,6 +28,14 @@ export const Overview = forwardRef(
...
+    const { fromNow, timeleftAsString } = useTimeLeft({
+      refreshInterval: 60,
+      refreshCallback: () => eraTimeLeft,
```
Comment (rossbulat) [inline src/modals/UnlockChunks/Overview.tsx:34]: Bit of a detour but can these be optional? Default `refreshInterval` to 60 and only call `refreshCallback` if it is provided.

**Review by rossbulat — COMMENTED** [review]
On the right track but some changes to be made. 

Noticed a bug where the time flickers every once in a while, and you are using `useEraTimeLeft` in the modal unnecessarily.

File: src/library/Hooks/useErasToTimeleft/index.tsx (hunk 3 above)
Comment (rossbulat) [reply src/library/Hooks/useErasToTimeleft/index.tsx:18]: I don't understand the question.

File: src/library/Hooks/useTimeLeft/types.ts [hunk 6]
```diff
@@ -28,6 +28,6 @@ export interface TimeLeftAll {
 }
 
 export interface TimeleftHookProps {
-  refreshCallback: () => number;
+  refreshCallback?: () => number;
```
Comment (rossbulat) [inline src/library/Hooks/useTimeLeft/types.ts:31]: remove ?

**Comment by rossbulat** [comment]
Screenshot 2023-01-04 at 15 23 17

Can you change this text to: "Unlocks can be withdrawn after 12 hours."

File: src/library/Hooks/useTimeLeft/index.tsx [hunk 7]
```diff
@@ -117,58 +114,63 @@ export const useTimeLeft = ({
...
-    //
-    // either handle regular timeleft update, or refresh `to` via `refreshCallback` every
-    // `refreshInterval` seconds.
-    const handleRefresh = () => {
-      if (r !== 0) {
-        setTimeleft(getTimeleft());
-      } else {
-        if (refreshCallback) {
-          setStateWithRef(fromNow(refreshCallback()), setTo, toRef);
+  if (TimeleftHook) {
```
Comment (rossbulat) [inline src/library/Hooks/useTimeLeft/index.tsx:117]: local variables should not be capitalised. Only globals should be.

File: src/library/Hooks/useTimeLeft/index.tsx [hunk 8]
```diff
@@ -117,58 +114,63 @@ export const useTimeLeft = ({
...
-        const interval = setInterval(() => {
-          if (inLastHour()) {
-            clearInterval(minIntervalRef.current);
-            setStateWithRef(undefined, setMinInterval, minIntervalRef);
-          }
-          r = Math.max(0, r - 60);
-          handleRefresh();
-        }, 60000);
-        setStateWithRef(interval, setMinInterval, minIntervalRef);
+          r = TimeleftHook.refreshInterval;
```
Comment (rossbulat) [inline src/library/Hooks/useTimeLeft/index.tsx:139]: This does not fall back to a default value.
`r  = timeleftHook?.refreshInterval || 60;`

can introduce defaults here too.

File: src/library/Hooks/useTimeLeft/index.tsx (hunk 7 above)
Comment (rossbulat) [inline src/library/Hooks/useTimeLeft/index.tsx:117]: useEffects should **Never** be scoped or wrapped in anything. They need to be top level

File: src/library/Hooks/useTimeLeft/index.tsx [hunk 9]
```diff
@@ -14,10 +14,7 @@ import {
   TimeLeftRaw,
 } from './types';
 
-export const useTimeLeft = ({
-  refreshCallback,
-  refreshInterval,
-}: TimeleftHookProps) => {
+export const useTimeLeft = (TimeleftHook?: TimeleftHookProps) => {
```
Comment (rossbulat) [inline src/library/Hooks/useTimeLeft/index.tsx]: `TmeleftHook` is not a hook, they can be treated as props to the hook. Therefore this should just be called `props`.

File: src/library/Hooks/useTimeLeft/index.tsx [hunk 10]
```diff
@@ -184,21 +186,31 @@ export const useTimeLeft = ({
   }, []);
 
   // format the duration as a string.
-  const timeleftAsString = (() => {
-    const { days, hours, minutes, seconds } = getDuration(toRef.current);
+  const timeleftAsString = (toDate?: Date, full?: boolean) => {
```
Comment (rossbulat) [inline src/library/Hooks/useTimeLeft/index.tsx:188]: `toDate` can just be called `to`

File: src/library/Hooks/useTimeLeft/index.tsx [hunk 11]
```diff
@@ -184,21 +186,31 @@ export const useTimeLeft = ({
...
+    const { days, hours, minutes, seconds } = toDate
+      ? getDuration(toDate)
+      : getDuration(toRef.current);
```
Comment (rossbulat) [inline src/library/Hooks/useTimeLeft/index.tsx:191]:
```suggestion
    const { days, hours, minutes, seconds } = getDuration(toDate || toRef.current);
```

File: src/library/Hooks/useTimeLeft/index.tsx [hunk 12]
```diff
@@ -184,21 +186,31 @@ export const useTimeLeft = ({
...
+    if (hours > 0) {
+      str += full
+        ? `${hours} ${t('time.hour', { count: hours, ns: 'base' })} `
+        : `${hours} ${t('time.hr', { count: hours, ns: 'base' })} `;
+    }
+    if (minutes > 0) {
+      str += full
+        ? `${minutes} ${t('time.minute', { count: minutes, ns: 'base' })}`
+        : `${minutes} ${t('time.min', { count: minutes, ns: 'base' })}`;
+    }
```
Comment (rossbulat) [inline src/library/Hooks/useTimeLeft/index.tsx:206]:
```suggestion
    let tHour = full ? 'time.hour' : 'time.hr';
    let tMinute = full ? 'time.minute' : 'time.min';
    
    if (hours > 0) {
      str += `${hours} ${t(tHour, { count: hours, ns: 'base' })} `
    }
    if (minutes > 0) {
      str +=
        `${minutes} ${t(tMinute, { count: minutes, ns: 'base' })}`
    }
```

File: src/library/StatBoxList/Timeleft.tsx [hunk 13]
```diff
@@ -6,7 +6,7 @@ import { StatPie } from 'library/Graphs/StatBoxPie';
 import { OpenHelpIcon } from 'library/OpenHelpIcon';
 import { StatBox } from './Item';
 import { TimeleftProps } from './types';
-import { TineleftWrapper } from './Wrapper';
+import { TimeleftWrapper } from './Wrapper';
```
Comment (rossbulat) [inline src/library/StatBoxList/Timeleft.tsx:9]: Let's make sure `t` is capitalised in timeleft. Rename to `TimeLeftWrapper`.

File: src/library/Hooks/useErasToTimeleft/index.tsx [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+import { useApi } from 'contexts/Api';
+
+export const useErasToTimeleft = () => {
+  const { consts } = useApi();
+  const { epochDuration, expectedBlockTime, sessionsPerEra } = consts;
+
+  // converts a number of eras to timeleft in seconds.
+  const erasToTimeLeft = (eras: number) => {
```
Comment (rossbulat) [inline src/library/Hooks/useErasToTimeleft/index.tsx:11]: Rename this function to `getErasToTimeLeft`.

When we are getting something (as opposed to setting something), we can use the `get` prefix.

**Comment by TingALin** [comment]
> Unlocks can be withdrawn after 12 hours.

done

**Review by rossbulat — APPROVED** [review]

**Review by rossbulat — APPROVED** [review]

**Pull request merged**